vault/resource_ad_secret_backend.go

```diff
@@ -0,0 +1,654 @@
+package vault
+
+import (
+	"fmt"
+	"github.com/terraform-providers/terraform-provider-vault/util"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/vault/api"
+)
+
+func adSecretBackendResource() *schema.Resource {
+	fields := map[string]*schema.Schema{
+		"backend": {
+			Type:        schema.TypeString,
+			Default:     "ad",
+			ForceNew:    true,
+			Optional:    true,
+			Description: `The mount path for a backend, for example, the path given in "$ vault auth enable -path=my-ad ad".`,
+			StateFunc: func(v interface{}) string {
+				return strings.Trim(v.(string), "/")
+			},
+		},
+		"anonymous_group_search": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: `Use anonymous binds when performing LDAP group searches (if true the initial credentials will still be used for the initial connection test).`,
+		},
+		"binddn": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: `Distinguished name of object to bind when performing user and group search.`,
+		},
+		"bindpass": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Sensitive:   true,
+			Description: `LDAP password for searching for the user DN.`,
+		},
+		"case_sensitive_names": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: `If true, case sensitivity will be used when comparing usernames and groups for matching policies.`,
+		},
+		"certificate": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: `CA certificate to use when verifying LDAP server certificate, must be x509 PEM encoded.`,
+		},
+		"client_tls_cert": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Sensitive:   true,
+			Description: `Client certificate to provide to the LDAP server, must be x509 PEM encoded.`,
+		},
+		"client_tls_key": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Sensitive:   true,
+			Description: `Client certificate key to provide to the LDAP server, must be x509 PEM encoded.`,
+		},
+		"default_lease_ttl_seconds": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: "Default lease duration for secrets in seconds",
+		},
+		"deny_null_bind": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: `Denies an unauthenticated LDAP bind request if the user's password is empty; defaults to true`,
+		},
+		"description": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Human-friendly description of the mount for the backend.",
+		},
+		"discoverdn": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: `Use anonymous bind to discover the bind DN of a user.`,
+		},
+		"formatter": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Deprecated:  `Formatter is deprecated and password_policy should be used.`,
```

Comment:
May want to mention the version for deprecation too.

```suggestion
			Deprecated:  `Formatter is deprecated and password_policy should be used with Vault >= 1.5.`,
```